packages/viz/src/lib/Layer.ts

```
@@ -1,107 +1,107 @@
 
-    this._layerStyles = new Style(style);
+    this._styles = new Style(style);
 
-    if (this._mvtLayerInstance) {
+    if (this._deckLayer) {
       await this._replaceLayer(previousSource);
     }
   }
 
+  /**
+   * Add the current layer to a Deck instance
+   * @param deckInstance instance to add the layer to
+   */
   public async addTo(deckInstance: any) {
+    // get current layers
     const currentDeckLayers = deckInstance.props.layers;
-    const createdDeckLayer = await this.getDeckGLLayer();
+    // create the new layer
+    const layer = await this._createDeckGLLayer();
 
     deckInstance.setProps({
       layers: [
         ...currentDeckLayers,
-        createdDeckLayer
+        layer
       ]
     });
 
     this._deckInstance = deckInstance;
   }
 
-  public async getDeckGLLayer() {
-    const layerProperties = await this._getDeckGLLayerProperties();
-
-    this._mvtLayerInstance = new MVTLayer(layerProperties);
-    return this._mvtLayerInstance;
-  }
+  /**
+   * Method to create the Deck.gl layer
+   */
+  private async _createDeckGLLayer() {
 
-  private async _getDeckGLLayerProperties() {
-    const { urlTemplates: data, geometryType } = await this._layerInstantiation.then(this._parseInstantiationResult);
-    const defaultGeometryStyles = defaultStyles[geometryType];
-    const layerId = this.generateLayerId(this._layerSource.getSourceValue());
+    // Get the blueprint of the layer
+    const blueprint = await this._source.blueprint();
 
-    return Object.assign(
-      { data, id: layerId },
-      defaultGeometryStyles.getProperties(),
-      this._layerStyles.getProperties()
+    // Build the properties expected by Deck.gl
+    const layerProperties =  Object.assign(
+      { data: blueprint.tileURL, id: this._source.id },
+      defaultStyles[blueprint.geometryType].getProperties(),
+      this._styles.getProperties()
     );
+
+    // Create the Deck.gl instance
+    if (this._source instanceof CARTOSource) {
```

I think we should check this at the beginning of the method and fail early without creating the layer, like:

```js
import { CartoError } from '@carto/toolkit-core';

private async _createDeckGLLayer() {
  if (!(this._source instanceof CARTOSource)) {
    throw new CartoError('Unsupported source instance')
  }

  // create layer and so on
}
```

---

The type of the source already forces it. The type is string | Source

At the constructor, we force to convert from string to CARTOSource, so at this point, the instance of the class is guaranteed to be a source